File: java/counting_valleys/CountingValleys.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;
import java.util.regex.*;
import java.util.stream.*;
import static java.util.stream.Collectors.joining;
import static java.util.stream.Collectors.toList;

public class CountingValleys {
    public static int countingValleys(int steps, String path) {
    // Write your code here
        int valleys = 0;
        int level = 0;
        int lastLevel = 0;
        char step;
        Map<Integer, Integer> socks = new HashMap<>();
        for (int i = 0; i < path.length(); i++) {
            step = path.charAt(i);
            lastLevel = level;
            if (step == 'D') {
                level--;
            } else {
                level++;
            }
            if (lastLevel < 0 && level == 0) {
                valleys++;
            }
        }
        return valleys;
    }

    public static void main(String[] args) {
        int steps = 8;
        String path = new String("UDDDUDUU");
        int result = countingValleys(steps, path);
        System.out.println(result);
    }
}

File: cpp/counting_valleys/counting_valleys.cpp
#include <bits/stdc++.h>

using namespace std;

int countingValleys(int steps, string path) {
  int valleys = 0;
  int level = 0;
  int lastLevel = 0;
  for (int i = 0; i < path.size(); i++) {
    char step = path[i];
    lastLevel = level;
    if(step == 'D') {
      level--;
    } else {
      level++;
    }
    if (lastLevel < 0 && level == 0) {
      valleys++;
    }
  }
  return valleys;
}

int main() {
  const int steps = 8;
  string path = "UDDDUDUU";
  const int valleys = countingValleys(steps, path);
  cout << valleys << '\n';
  return 0;
}